File: ios/StockbookTests/StatementTests.swift
import Testing
import Foundation
@testable import Stockbook

/// The date arithmetic behind "this month" and "this year".
///
/// Pinned to a fixed UTC Gregorian calendar rather than the device's, because a
/// month boundary is exactly the kind of thing that passes in London and fails in
/// Riyadh — and this app is for a shop in Saudi Arabia read on a phone that may be
/// set to anything.
@Suite("Statement periods")
struct StatementPeriodTests {

    private var calendar: Calendar {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(secondsFromGMT: 0)!
        return calendar
    }

    private func date(_ year: Int, _ month: Int, _ day: Int, _ hour: Int = 12) -> Date {
        calendar.date(from: DateComponents(year: year, month: month, day: day, hour: hour))!
    }

    @Test("A month runs from its first instant to the next month's, exclusive")
    func monthRange() {
        let range = StatementPeriod.month(date(2026, 8, 17)).range(calendar: calendar)

        #expect(range.start == date(2026, 8, 1, 0))
        #expect(range.end == date(2026, 9, 1, 0))
        #expect(range.contains(date(2026, 8, 1, 0)))
        #expect(range.contains(date(2026, 8, 31, 23)))
        // The load-bearing one. Midnight on the 1st belongs to September, and if
        // it belonged to both, a bill written then would be counted twice.
        #expect(!range.contains(date(2026, 9, 1, 0)))
        #expect(!range.contains(date(2026, 7, 31, 23)))
    }

    @Test("A year runs January to January")
    func yearRange() {
        let range = StatementPeriod.year(date(2026, 8, 17)).range(calendar: calendar)

        #expect(range.start == date(2026, 1, 1, 0))
        #expect(range.end == date(2027, 1, 1, 0))
        #expect(range.contains(date(2026, 12, 31, 23)))
        #expect(!range.contains(date(2027, 1, 1, 0)))
    }

    @Test("A chosen range covers both end days whole")
    func customRange() {
        let range = StatementPeriod
            .custom(from: date(2026, 8, 3, 15), to: date(2026, 8, 5, 9))
            .range(calendar: calendar)

        // Picked mid-afternoon on the 3rd, but the owner means the 3rd.
        #expect(range.start == date(2026, 8, 3, 0))
        #expect(range.contains(date(2026, 8, 3, 0)))
        // And the whole of the 5th, not up to 9am on it.
        #expect(range.contains(date(2026, 8, 5, 23)))
        #expect(!range.contains(date(2026, 8, 6, 0)))
    }

    @Test("A range picked backwards is still the range they meant")
    func customRangeReversed() {
        let forwards = StatementPeriod
            .custom(from: date(2026, 8, 3), to: date(2026, 8, 5))
            .range(calendar: calendar)
        let backwards = StatementPeriod
            .custom(from: date(2026, 8, 5), to: date(2026, 8, 3))
            .range(calendar: calendar)

        #expect(forwards == backwards)
    }

    @Test("Last month is last month, including from the 1st")
    func lastMonth() {
        let range = StatementPeriod
            .lastMonth(now: date(2026, 1, 1, 3), calendar: calendar)
            .range(calendar: calendar)

        // From the small hours of New Year's Day, "last month" is December.
        #expect(range.start == date(2025, 12, 1, 0))
        #expect(range.end == date(2026, 1, 1, 0))
    }
}

/// What a statement says, which is the whole feature.
@Suite("Statements")
struct StatementTests {

    private var calendar: Calendar {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(secondsFromGMT: 0)!
        return calendar
    }

    private func date(_ year: Int, _ month: Int, _ day: Int) -> Date {
        calendar.date(from: DateComponents(year: year, month: month, day: day, hour: 12))!
    }

    private let ahmed = Customer(
        name: "Ahmed Contracting",
        key: "ahmed contracting",
        billCount: 0,
        total: 0,
        owed: 0,
        phone: nil,
        place: nil,
        openingBalance: 0,
        isOnRoster: true
    )

    private func bill(
        _ number: Int,
        on date: Date,
        total: Double,
        paid: Double? = nil,
        voided: Bool = false
    ) -> Bill {
        Bill(
            number: number,
            lines: [BillLine(productUID: nil, name: "Cisa lock", qty: 1, price: total)],
            total: total,
            paid: paid,
            who: "Ahmed Contracting",
            createdAt: date,
            voided: voided
        )
    }

    private func payment(_ amount: Double, on date: Date) -> Payment {
        Payment(customerKey: "ahmed contracting", amount: amount, receivedAt: date)
    }

    private func august(
        bills: [Bill],
        payments: [Payment] = []
    ) -> Statement {
        Statement.make(
            customer: ahmed,
            bills: bills,
            payments: payments,
            period: .month(date(2026, 8, 10)),
            calendar: calendar
        )
    }

    @Test("A bill paid in full leaves nothing owed")
    func paidInFull() {
        let statement = august(bills: [bill(1, on: date(2026, 8, 4), total: 900)])

        #expect(statement.billed == 900)
        #expect(statement.received == 900)
        #expect(statement.closingBalance == 0)
        #expect(statement.openingBalance == 0)
    }

    @Test("A part payment at the counter leaves the rest owed")
    func partPaid() {
        let statement = august(bills: [bill(1, on: date(2026, 8, 4), total: 900, paid: 500)])

        #expect(statement.billed == 900)
        #expect(statement.received == 500)
        #expect(statement.closingBalance == 400)
    }

    /// The reason payments exist. Without them this figure could never come down.
    @Test("A payment received later clears the balance")
    func paymentClears() {
        let statement = august(
            bills: [bill(1, on: date(2026, 8, 4), total: 900, paid: 500)],
            payments: [payment(400, on: date(2026, 8, 20))]
        )

        #expect(statement.received == 900)
        #expect(statement.closingBalance == 0)
        #expect(statement.entries.count == 2)
    }

    /// The difference between a statement and a filtered list of bills.
    @Test("What was owed before the period is brought forward")
    func openingBalance() {
        let statement = august(
            bills: [
                bill(1, on: date(2026, 3, 2), total: 900, paid: 0),      // owed since March
                bill(2, on: date(2026, 8, 4), total: 100, paid: 0)
            ]
        )

        #expect(statement.openingBalance == 900)
        #expect(statement.billed == 100, "March is not in August's figures")
        #expect(statement.closingBalance == 1000)
        // Only August's bill is listed; the March one is carried as a figure.
        let numbers = statement.entries.compactMap { entry -> Int? in
            if case .bill(let bill) = entry { return bill.number }
            return nil
        }
        #expect(numbers == [2])
    }

    @Test("A payment before the period comes off the brought-forward figure")
    func paymentBeforePeriod() {
        let statement = august(
            bills: [bill(1, on: date(2026, 3, 2), total: 900, paid: 0)],
            payments: [payment(400, on: date(2026, 7, 30))]
        )

        #expect(statement.openingBalance == 500)
        #expect(statement.closingBalance == 500)
        #expect(statement.isEmpty, "nothing happened in August")
    }

    @Test("A voided bill is listed and counts for nothing")
    func voided() {
        let statement = august(
            bills: [
                bill(1, on: date(2026, 8, 4), total: 900, paid: 0),
                bill(2, on: date(2026, 8, 5), total: 500, paid: 0, voided: true)
            ]
        )

        #expect(statement.entries.count == 2, "history is marked, never hidden")
        #expect(statement.billed == 900)
        #expect(statement.closingBalance == 900)
    }

    @Test("Entries read downwards, oldest first")
    func order() {
        let statement = august(
            bills: [
                bill(2, on: date(2026, 8, 20), total: 100),
                bill(1, on: date(2026, 8, 4), total: 200)
            ],
            payments: [payment(50, on: date(2026, 8, 10))]
        )

        let dates = statement.entries.map(\.date)
        let sorted = dates.sorted()
        // Every other list in this app is newest-first. A statement is a
        // document, and a document reads down the page.
        #expect(dates == sorted)
    }

    /// An invariant, not an example: the two are computed by different routes —
    /// one accumulates per entry, the other is opening + billed − received — and
    /// a statement whose column disagrees with its own total is worthless.
    @Test("The running balance lands exactly on the closing balance")
    func runningBalanceAgrees() {
        let statement = august(
            bills: [
                bill(1, on: date(2026, 3, 2), total: 900, paid: 0),
                bill(2, on: date(2026, 8, 4), total: 250, paid: 100),
                bill(3, on: date(2026, 8, 6), total: 80),
                bill(4, on: date(2026, 8, 9), total: 400, paid: 0, voided: true)
            ],
            payments: [payment(200, on: date(2026, 8, 12)), payment(25, on: date(2026, 8, 28))]
        )

        #expect(statement.runningBalances.count == statement.entries.count)
        let last = statement.runningBalances.last
        #expect(last == statement.closingBalance)
    }

    @Test("Paying more than owed shows as an advance rather than a wrong total")
    func overpayment() {
        let statement = august(
            bills: [bill(1, on: date(2026, 8, 4), total: 100, paid: 0)],
            payments: [payment(250, on: date(2026, 8, 20))]
        )

        #expect(statement.closingBalance == -150)
        // And the customer row says it in words rather than showing a minus sign.
        let paidAhead = Customer(
            name: "Ahmed", key: "ahmed", billCount: 1, total: 100, owed: -150,
            phone: nil, place: nil, openingBalance: 0, isOnRoster: true
        )
        let meta = paidAhead.meta(in: .default, strings: Strings(language: .english))
        #expect(meta.contains("in advance"))
    }

    @Test("A customer with no history at all yields an empty statement, not nil")
    func nothingAtAll() {
        let statement = august(bills: [])

        #expect(statement.isEmpty)
        #expect(statement.openingBalance == 0)
        #expect(statement.closingBalance == 0)
        #expect(statement.runningBalances.isEmpty)
    }
}

File: android/core/src/test/kotlin/com/stockbook/core/StatementTests.kt
package com.stockbook.core

import com.stockbook.core.model.Bill
import com.stockbook.core.model.BillLine
import com.stockbook.core.model.Currency
import com.stockbook.core.model.Customer
import com.stockbook.core.model.Payment
import com.stockbook.core.model.Statement
import com.stockbook.core.model.StatementPeriod
import com.stockbook.core.text.AppLanguage
import com.stockbook.core.text.Strings
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.ZoneId
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

/**
 * The date arithmetic behind "this month" and "this year".
 *
 * Pinned to a fixed UTC zone rather than the device's, because a month boundary
 * is exactly the kind of thing that passes in London and fails in Riyadh — and
 * this app is for a shop in Saudi Arabia read on a phone that may be set to
 * anything.
 */
class StatementPeriodTests {

    private val utc: ZoneId = ZoneId.of("UTC")

    private fun at(year: Int, month: Int, day: Int, hour: Int = 12): Instant =
        LocalDateTime.of(year, month, day, hour, 0).atZone(utc).toInstant()

    @Test
    fun `a month runs from its first instant to the next month's exclusive`() {
        val range = StatementPeriod.Month(at(2026, 8, 17)).range(utc)

        assertEquals(at(2026, 8, 1, 0), range.start)
        assertEquals(at(2026, 9, 1, 0), range.end)
        assertTrue(at(2026, 8, 1, 0) in range)
        assertTrue(at(2026, 8, 31, 23) in range)
        // The load-bearing one. Midnight on the 1st belongs to September, and if
        // it belonged to both, a bill written then would be counted twice.
        assertFalse(at(2026, 9, 1, 0) in range)
        assertFalse(at(2026, 7, 31, 23) in range)
    }

    @Test
    fun `a year runs January to January`() {
        val range = StatementPeriod.Year(at(2026, 8, 17)).range(utc)

        assertEquals(at(2026, 1, 1, 0), range.start)
        assertEquals(at(2027, 1, 1, 0), range.end)
        assertTrue(at(2026, 12, 31, 23) in range)
        assertFalse(at(2027, 1, 1, 0) in range)
    }

    @Test
    fun `a chosen range covers both end days whole`() {
        val range = StatementPeriod.Custom(at(2026, 8, 3, 15), at(2026, 8, 5, 9)).range(utc)

        // Picked mid-afternoon on the 3rd, but the owner means the 3rd.
        assertEquals(at(2026, 8, 3, 0), range.start)
        assertTrue(at(2026, 8, 3, 0) in range)
        // And the whole of the 5th, not up to 9am on it.
        assertTrue(at(2026, 8, 5, 23) in range)
        assertFalse(at(2026, 8, 6, 0) in range)
    }

    @Test
    fun `a range picked backwards is still the range they meant`() {
        val forwards = StatementPeriod.Custom(at(2026, 8, 3), at(2026, 8, 5)).range(utc)
        val backwards = StatementPeriod.Custom(at(2026, 8, 5), at(2026, 8, 3)).range(utc)

        assertEquals(forwards, backwards)
    }

    @Test
    fun `last month is last month including from the first`() {
        val range = StatementPeriod.lastMonth(at(2026, 1, 1, 3), utc).range(utc)

        // From the small hours of New Year's Day, "last month" is December.
        assertEquals(at(2025, 12, 1, 0), range.start)
        assertEquals(at(2026, 1, 1, 0), range.end)
    }

    @Test
    fun `a month is found the same way whichever day of it you are on`() {
        val early = StatementPeriod.Month(at(2026, 8, 1, 0)).range(utc)
        val late = StatementPeriod.Month(at(2026, 8, 31, 23)).range(utc)
        assertEquals(early, late)
    }
}

/** What a statement says, which is the whole feature. */
class StatementTests {

    private val utc: ZoneId = ZoneId.of("UTC")

    private fun at(year: Int, month: Int, day: Int): Instant =
        LocalDate.of(year, month, day).atTime(12, 0).atZone(utc).toInstant()

    private val ahmed = Customer(
        name = "Ahmed Contracting",
        key = "ahmed contracting",
        billCount = 0,
        total = 0.0,
        owed = 0.0,
        isOnRoster = true
    )

    private fun bill(
        number: Int,
        on: Instant,
        total: Double,
        paid: Double? = null
    ) = Bill(
        number = number,
        lines = listOf(BillLine(productUid = null, name = "Cisa lock", qty = 1, price = total)),
        total = total,
        paid = paid,
        who = "Ahmed Contracting",
        createdAt = on
    )

    private fun payment(amount: Double, on: Instant) =
        Payment(customerKey = "ahmed contracting", amount = amount, receivedAt = on)

    private fun august(bills: List<Bill>, payments: List<Payment> = emptyList()): Statement =
        Statement.make(
            customer = ahmed,
            bills = bills,
            payments = payments,
            period = StatementPeriod.Month(at(2026, 8, 10)),
            zone = utc
        )

    @Test
    fun `a bill paid in full leaves nothing owed`() {
        val statement = august(listOf(bill(1, at(2026, 8, 4), 900.0)))

        assertEquals(900.0, statement.billed)
        assertEquals(900.0, statement.received)
        assertEquals(0.0, statement.closingBalance)
        assertEquals(0.0, statement.openingBalance)
    }

    @Test
    fun `a part payment at the counter leaves the rest owed`() {
        val statement = august(listOf(bill(1, at(2026, 8, 4), 900.0, paid = 500.0)))

        assertEquals(900.0, statement.billed)
        assertEquals(500.0, statement.received)
        assertEquals(400.0, statement.closingBalance)
    }

    /** The reason payments exist. Without them this figure could never come down. */
    @Test
    fun `a payment received later clears the balance`() {
        val statement = august(
            listOf(bill(1, at(2026, 8, 4), 900.0, paid = 500.0)),
            listOf(payment(400.0, at(2026, 8, 20)))
        )

        assertEquals(900.0, statement.received)
        assertEquals(0.0, statement.closingBalance)
        assertEquals(2, statement.entries.size)
    }

    /** The difference between a statement and a filtered list of bills. */
    @Test
    fun `what was owed before the period is brought forward`() {
        val statement = august(
            listOf(
                bill(1, at(2026, 3, 2), 900.0, paid = 0.0),
                bill(2, at(2026, 8, 4), 100.0, paid = 0.0)
            )
        )

        assertEquals(900.0, statement.openingBalance)
        assertEquals(100.0, statement.billed, "March is not in August's figures")
        assertEquals(1000.0, statement.closingBalance)
        val numbers = statement.entries.mapNotNull { (it as? Statement.Entry.ForBill)?.bill?.number }
        assertEquals(listOf(2), numbers)
    }

    @Test
    fun `a payment before the period comes off the brought-forward figure`() {
        val statement = august(
            listOf(bill(1, at(2026, 3, 2), 900.0, paid = 0.0)),
            listOf(payment(400.0, at(2026, 7, 30)))
        )

        assertEquals(500.0, statement.openingBalance)
        assertEquals(500.0, statement.closingBalance)
        assertTrue(statement.isEmpty, "nothing happened in August")
    }

    @Test
    fun `entries read downwards oldest first`() {
        val statement = august(
            listOf(bill(2, at(2026, 8, 20), 100.0), bill(1, at(2026, 8, 4), 200.0)),
            listOf(payment(50.0, at(2026, 8, 10)))
        )

        val dates = statement.entries.map { it.date }
        // Every other list in this app is newest-first. A statement is a
        // document, and a document reads down the page.
        assertEquals(dates.sorted(), dates)
    }

    /**
     * An invariant, not an example: the two are computed by different routes —
     * one accumulates per entry, the other is opening + billed − received — and a
     * statement whose column disagrees with its own total is worthless.
     */
    @Test
    fun `the running balance lands exactly on the closing balance`() {
        val statement = august(
            listOf(
                bill(1, at(2026, 3, 2), 900.0, paid = 0.0),
                bill(2, at(2026, 8, 4), 250.0, paid = 100.0),
                bill(3, at(2026, 8, 6), 80.0),
                bill(4, at(2026, 8, 9), 400.0, paid = 0.0)
            ),
            listOf(payment(200.0, at(2026, 8, 12)), payment(25.0, at(2026, 8, 28)))
        )

        assertEquals(statement.entries.size, statement.runningBalances.size)
        assertEquals(statement.closingBalance, statement.runningBalances.last())
    }

    @Test
    fun `paying more than owed shows as an advance rather than a wrong total`() {
        val statement = august(
            listOf(bill(1, at(2026, 8, 4), 100.0, paid = 0.0)),
            listOf(payment(250.0, at(2026, 8, 20)))
        )

        assertEquals(-150.0, statement.closingBalance)
        // And the customer row says it in words rather than showing a minus sign.
        val paidAhead = ahmed.copy(billCount = 1, total = 100.0, owed = -150.0)
        val meta = paidAhead.meta(Currency.default, Strings(AppLanguage.ENGLISH))
        assertTrue(meta.contains("in advance"), meta)
    }

    @Test
    fun `a customer with no history at all yields an empty statement`() {
        val statement = august(emptyList())

        assertTrue(statement.isEmpty)
        assertEquals(0.0, statement.openingBalance)
        assertEquals(0.0, statement.closingBalance)
        assertTrue(statement.runningBalances.isEmpty())
    }
}
